spec/PasswordPolicy.spec.js

```diff
@@ -757,4 +759,233 @@ describe("Password Policy: ", () => {
+    }).then(() => {
+      fail('passwordPolicy.daysBeforeExpiry negative number test failed');
+      done();
+    }).catch(err => {
+      expect(err).toEqual('passwordPolicy.daysBeforeExpiry must be a positive number');
+      done();
+    });
+  });
+
+  it('should succeed if logged in before password expires', (done) => {
+    const user = new Parse.User();
+    reconfigureServer({
+      appName: 'passwordPolicy',
+      passwordPolicy: {
+        daysBeforeExpiry: 1 // 1 day
+      },
+      publicServerURL: "http://localhost:8378/1"
+    }).then(() => {
+      user.setUsername("user1");
+      user.setPassword("user1");
+      user.set('email', 'user1@parse.com');
+      user.signUp().then((u) => {
+        Parse.User.logIn("user1", "user1").then((user) => {
+          done();
+        }).catch((error) => {
+          jfail(error);
+          fail('Login should have succeeded before password expiry.');
+          done();
+        });
+      }).catch((error) => {
+        jfail(error);
+        fail('Signup failed.');
+        done();
+      });
+    })
+  });
+
+  it('should fail if logged in after password expires', (done) => {
+    const user = new Parse.User();
+    reconfigureServer({
+      appName: 'passwordPolicy',
+      passwordPolicy: {
+        daysBeforeExpiry: 0.000005787 // 0.5 sec
+      },
+      publicServerURL: "http://localhost:8378/1"
+    }).then(() => {
+      user.setUsername("user1");
+      user.setPassword("user1");
+      user.set('email', 'user1@parse.com');
+      user.signUp().then(() => {
+        // wait for a bit more than the validity duration set
+        setTimeout(() => {
+          Parse.User.logIn("user1", "user1").then(() => {
+            fail("logIn should have failed");
+            done();
+          }).catch((error) => {
+            expect(error.code).toEqual(Parse.Error.OBJECT_NOT_FOUND);
+            expect(error.message).toEqual('Your password has expired. Please reset your password.');
+            done();
+          });
+        }, 1000);
+      }).catch((error) => {
+        jfail(error);
+        fail('Signup failed.');
+        done();
+      });
+    });
+  });
+
+  it('should apply password expiry policy to existing user upon first login after policy is enabled', (done) => {
+    const user = new Parse.User();
+    reconfigureServer({
+      appName: 'passwordPolicy',
+      publicServerURL: "http://localhost:8378/1"
+    }).then(() => {
+      user.setUsername("user1");
+      user.setPassword("user1");
+      user.set('email', 'user1@parse.com');
+      user.signUp().then(() => {
+        Parse.User.logOut();
+        reconfigureServer({
+          appName: 'passwordPolicy',
+          passwordPolicy: {
+            daysBeforeExpiry: 0.000005787 // 0.5 sec
+          },
+          publicServerURL: "http://localhost:8378/1"
+        }).then(() => {
+          Parse.User.logIn("user1", "user1").then((u) => {
+            Parse.User.logOut();
```

Comment:
I agree chaining is better practice in general. It also allows us to handle errors at one place. However it may be better to handle separately in cases where the test cases expect a mix of SUCCESS and FAILURE resolutions in order to display appropriate messages.

I will take a look again and refactor the pieces that make sense.